### PR TITLE
Aho-Corasick (Issue #2118)

### LINKED ISSUE #2118: Aho-Corasick
- [ ] C - @santushtisharma10
- [x] C++
- [ ] C#
- [ ] CoffeeScript
- [ ] Dart
- [ ] Go
- [ ] Java - @topcoder69
- [ ] JS - @kashika3101
- [ ] Kotlin
- [ ] Php
- [x] Python - @shivanidalmia1203
- [ ] Readme - @Raksha25
- [ ] Ruby
- [ ] Rust
- [ ] TypeScript

### PR DESCRIPTION
**Pull request by Anushka118**
Fixes #2118

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- Implemented Aho-Corasick using Python

#### Languages Used:
- Python 

#### Files Added:
- AhoCorasick.py 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by asha15** [comment]
please specify the issue number in PR template

**Comment by Anushka118** [comment]
specified the issue number in PR template

**Comment by Anushka118** [comment]
@asha15 i did the necessary changes please merge the PR as i am not able to work on further issues

File: AhoCorasick.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
+# implementing AhoCorasick using adjacency list
+import collections
```
Comment (asha15) [inline AhoCorasick.py:2]: please include this file in the relevant folder 

rename the file name as Aho_Corasick.py

File: AhoCorasick.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
+# implementing AhoCorasick using adjacency list
```
Comment (asha15) [inline AhoCorasick.py:1]: please give a brief description about the algorithm. 
- max 4 lines
- one line should contain max 80 characters

File: AhoCorasick.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
+# implementing AhoCorasick using adjacency list
+import collections
+
+AList = []  # declaring thr adjacency list
```
Comment (asha15) [inline AhoCorasick.py:4]: please correct the typo in the

File: AhoCorasick.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
+# implementing AhoCorasick using adjacency list
+import collections
+
+AList = []  # declaring thr adjacency list
+
```
Comment (asha15) [inline AhoCorasick.py:5]: please remove this line

File: AhoCorasick.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+def init_trie(wordList):
+    createTrie()
+    addKeyWords(wordList)
+    setFailTrans()
+
+
+def createTrie():  # initializing a root of the trie
+
```
Comment (asha15) [inline AhoCorasick.py:14]: remove this line

File: AhoCorasick.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    AList.append({'character': ' ', 'nextState': [], 'failedState': 0, 'outputList': []})
+
```
Comment (asha15) [inline AhoCorasick.py:16]: remove this line. 
please remove the extra line in your code

File: AhoCorasick.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+# character -> the character the root node is repesenting
+# nextStates -> list of ids of child nodes of the root node
+# failedState -> id of the failed state
+# outputList -> list of all the valid words encountered
+
+def addKeyWords(wordList):  # ading all the valid words in list of valid words
+    for validWord in wordList:
+        addWordList(validWord)
+
```
Comment (asha15) [inline AhoCorasick.py:26]: remove this line

File: AhoCorasick.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+def findNextState(currState, val):
+    # takes a node and value and return the id of the child of that node
+    # whose character matches value or else return none if not found
+    for node in AList[currState]["nextState"]:
+        if AList[node]["character"] == val:
+            return node
+
+    return None
+
```
Comment (asha15) [inline AhoCorasick.py:36]: remove the extra lines

File: AhoCorasick.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        # creating a new node for a character
+        node = {'character': keyword[i], 'nextState': [], 'failedState': 0, 'outputList': []}
+
+        # appending the character to the Adjacency List
+        AList.append(node)
+        AList[currState]["nextState"].append(len(AList) - 1)
+        currState = len(AList) - 1
+    # appending te keyword to the output of the last node
+    AList[currState]["outputList"].append(keyword)
+
```
Comment (asha15) [inline AhoCorasick.py:63]: remove this line

File: AhoCorasick.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+# for fail transitions, breadth first search will be done on the trie
+
```
Comment (asha15) [inline AhoCorasick.py:66]: remove this line

File: AhoCorasick.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+            if AList[child]["failedState"] is None:
+                AList[child]["failedState"] = 0
+
+            AList[child]["outputList"] = AList[child]["outputList"] + AList[AList[child]["failedState"]]["outputList"]
+
+
+def GetKeywords(words):
+    currState = 0
+    for i in range(len(words)):
+
```
Comment (asha15) [inline AhoCorasick.py:93]: remove this line

File: AhoCorasick.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        if currState is None:
+            currState = 0
+
+        else:
+            for j in AList[currState]["outputList"]:
+                start = i - len(j) + 1
+                end = len(j) + start -1
+                print("keyword", j, " appears from ",start , " to " ,end)
+
```
Comment (asha15) [inline AhoCorasick.py:107]: remove this line

**Comment by Anushka118** [comment]
did the necessary changes. Please merge the file so that i can work on further issues.

**Comment by asha15** [comment]
please remove the irrelevant files from the PR, squash your commits and let me know

**Comment by somya-kapoor** [comment]
Are you working on it?

**Comment by somya-kapoor** [comment]
Due to no activity for more than 7 days, I am closing this issue.

**Comment by Anushka118** [comment]
i have made the necessary changes in this issue... please merge it

**Comment by asha15** [comment]
you have added three files to this PR. please create a new PR and tag me

**Pull request closed without merging**